src/modules/flow/dcmi.c

```diff
@@ -251,35 +251,47 @@ void dma_copy_image_buffers(uint8_t ** current_image, uint8_t ** previous_image,
 	*current_image = *previous_image;
 	*previous_image = tmp_image;
 
-TODO(NB dma_copy_image_buffers is calling uavcan_run());
-
 	/* wait for new image if needed */
-	while(image_counter < image_step) {
-            PROBE_1(false);
-            uavcan_run();
-            PROBE_1(true);
-	}
+	while(image_counter < image_step);
 
 	image_counter = 0;
 
 	/* time between images */
 	time_between_images = time_between_next_images;
 
+	uint8_t *source = NULL;
+
 	/* copy image */
-	if (dcmi_image_buffer_unused == 1)
-	{
-		for (uint16_t pixel = 0; pixel < image_size; pixel++)
-			(*current_image)[pixel] = (uint8_t)(dcmi_image_buffer_8bit_1[pixel]);
+	if (dcmi_image_buffer_unused == 1) {
+		source = dcmi_image_buffer_8bit_1;
+	} else if (dcmi_image_buffer_unused == 2) {
+		source = dcmi_image_buffer_8bit_2;
+	} else {
+		source = dcmi_image_buffer_8bit_3;
 	}
-	else if (dcmi_image_buffer_unused == 2)
-	{
-		for (uint16_t pixel = 0; pixel < image_size; pixel++)
-			(*current_image)[pixel] = (uint8_t)(dcmi_image_buffer_8bit_2[pixel]);
-	}
-	else
-	{
-		for (uint16_t pixel = 0; pixel < image_size; pixel++)
-			(*current_image)[pixel] = (uint8_t)(dcmi_image_buffer_8bit_3[pixel]);
+
+	for (uint16_t pixel = 0; pixel < image_size; pixel++)
+		(*current_image)[pixel] = source[pixel];
+}
+
+void whitened_image(uint8_t *source, uint8_t *dest, uint16_t image_size) {
+	double sum = 0.0;
+	for (uint16_t pixel = 0; pixel < image_size; pixel++)
+		sum += source[pixel];
+	double mean = sum / image_size;
+	double rss = 0.0;
+	for (uint16_t pixel = 0; pixel < image_size; pixel++)
+		rss += pow(source[pixel] - mean, 2);
```

Comment:
This executes a double instruction and pow instead of square. Please replace with: `(source[pixel] - mean) * (source[pixel] - mean)`. This will be orders of magnitude faster.